subsys/bluetooth/services/ots_olcp.c

```diff
@@ -0,0 +1,313 @@
+}
+
+static enum bt_gatt_ots_olcp_res_code olcp_proc_execute(
+	struct bt_gatt_ots *ots, struct bt_gatt_ots_olcp_proc *proc)
+{
+	LOG_DBG("Executing OLCP procedure with 0x%02X Op Code", proc->type);
+
+	switch (proc->type) {
+	case BT_GATT_OTS_OLCP_PROC_FIRST:
+		return olcp_first_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_LAST:
+		return olcp_last_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_PREV:
+		return olcp_prev_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_NEXT:
+		return olcp_next_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_GOTO:
+		return olcp_goto_proc_execute(ots, proc->goto_params.id);
+	case BT_GATT_OTS_OLCP_PROC_ORDER:
+	case BT_GATT_OTS_OLCP_PROC_REQ_NUM_OBJS:
+	case BT_GATT_OTS_OLCP_PROC_CLEAR_MARKING:
+	default:
+		return BT_GATT_OTS_OLCP_RES_PROC_NOT_SUP;
+	}
+};
+
+static enum bt_gatt_ots_olcp_res_code olcp_command_decode(
+	const uint8_t *buf, struct bt_gatt_ots_olcp_proc *proc)
+{
+	memset(proc, 0, sizeof(*proc));
+
+	proc->type = *buf++;
+
+	switch (proc->type) {
+	case BT_GATT_OTS_OLCP_PROC_FIRST:
+	case BT_GATT_OTS_OLCP_PROC_LAST:
+	case BT_GATT_OTS_OLCP_PROC_PREV:
+	case BT_GATT_OTS_OLCP_PROC_NEXT:
+		break;
+	case BT_GATT_OTS_OLCP_PROC_GOTO:
+	{
+		struct bt_gatt_ots_obj_id *id;
+
+		id = (struct bt_gatt_ots_obj_id *) &proc->goto_params.id;
+		memcpy(id->ba, buf, sizeof(id->ba));
+		break;
+	}
+	default:
+		LOG_WRN("OLCP unsupported procedure type");
+		return BT_GATT_OTS_OLCP_RES_PROC_NOT_SUP;
+	}
+
+	return BT_GATT_OTS_OLCP_RES_SUCCESS;
+}
+
+static bool olcp_command_len_verify(enum bt_gatt_ots_olcp_proc_type type,
+				    uint16_t len)
+{
+	uint16_t ref_len = OLCP_PROC_TYPE_SIZE;
+
+	switch (type) {
+	case BT_GATT_OTS_OLCP_PROC_FIRST:
+	case BT_GATT_OTS_OLCP_PROC_LAST:
+	case BT_GATT_OTS_OLCP_PROC_PREV:
+	case BT_GATT_OTS_OLCP_PROC_NEXT:
+	case BT_GATT_OTS_OLCP_PROC_REQ_NUM_OBJS:
+	case BT_GATT_OTS_OLCP_PROC_CLEAR_MARKING:
+		break;
+	case BT_GATT_OTS_OLCP_PROC_GOTO:
+		ref_len += sizeof(((struct bt_gatt_ots_obj_id *) 0)->ba);
+		break;
+	case BT_GATT_OTS_OLCP_PROC_ORDER:
+		ref_len += sizeof(uint8_t);
+		break;
+	default:
+		return true;
+	}
+
+	return (len == ref_len);
+}
+
+static void olcp_ind_cb(struct bt_conn *conn,
+		const struct bt_gatt_attr *attr,
+		uint8_t err)
+{
+	LOG_DBG("Received OLCP Indication ACK with status: 0x%04X", err);
+}
+
+static int olcp_ind_send(const struct bt_gatt_attr *olcp_attr,
+			 enum bt_gatt_ots_olcp_proc_type req_op_code,
+			 enum bt_gatt_ots_olcp_res_code olcp_status)
+{
+	uint8_t olcp_res[OLCP_RES_MAX_SIZE];
+	uint16_t olcp_res_len = 0;
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) olcp_attr->user_data;
+
+	/* Encode OLCP Response */
+	olcp_res[olcp_res_len++] = BT_GATT_OTS_OLCP_PROC_RESP;
+	olcp_res[olcp_res_len++] = req_op_code;
+	olcp_res[olcp_res_len++] = olcp_status;
+
+	/* Prepare indication parameters */
+	memset(&ots->olcp_ind.params, 0, sizeof(ots->olcp_ind.params));
+	memcpy(&ots->olcp_ind.attr, olcp_attr, sizeof(ots->olcp_ind.attr));
```

Comment:
This is just a pointer, isn't is not? I guess a simple assign would give the same result?

Reply:
No, it's a structure holding attribute content. I copy it here because I found no other way to pass OTS instance context to the indication callback.

Reply:
Because `BT_GATT_CCC_MANAGED` doens't have support for user_data? 

But if `olcp_attr` is the write attribute, which does have it as you do `struct bt_gatt_ots *ots = (struct bt_gatt_ots *) oacp_attr->user_data;`, wouldn't `ots->oacp_ind.params.attr = olcp_attr;` give the same result? Or am I missing something here?

Reply:
yes, I tried doing it as you said but it did not work, unfortunately. Upon further investigation, it turned out that ``olcp_attr`` points to an automatic variable which becomes invalid (overwritten memory by other variables) in the indication callback context.

Reply:
I see; that's rather surprising. But then yes, doing this copy would be a way of getting the OTS content. 

I've struggled with the same challenge, where my solution had been to iterate over all the attributes of all the service instances to find the pointer that matches the pointer in the indication callback, but this is likely more efficient just by storing some data.

Reply:
Couldn't we use CONTAINER_OF in some way to get around this? I don't remember all the details but I think we solved this problem for HIDS using BT_GATT_CCC_INITIALIZER (although it goes through the gatt_pool.

Reply:
CONTAINER_OF is used in the indication callback. However, it works only if the attribute data is in static memory.